confgenerator/resourcedetector/gce_detector.go

```diff
@@ -0,0 +1,128 @@
+package resourcedetector
+
+type gceAttribute int
+
+const (
+	project gceAttribute = iota
+	zone
+	network
+	subnetwork
+	publicIP
+	privateIP
+	instanceID
+	instanceName
+	tags
+	machineType
+	metadata
+	label
+	interfaceIPv4
+)
+
+func GetGCEDetector() (Detector, error) {
+	provider := NewGCEMetadataProvider()
+	dt := GCEDetectorBuilder{provider: provider}
+	return dt.GetDetector()
+}
+
+// The data provider interface for GCE environment
+// Implementation of this provider can use either the metadata server on VM,
+// or the cloud API
+type gceDataProvider interface {
+	getProject() (string, error)
+	getZone() (string, error)
+	getNetwork() (string, error)
+	getSubnetwork() (string, error)
+	getPublicIP() (string, error)
+	getPrivateIP() (string, error)
+	getInstanceID() (string, error)
+	getInstanceName() (string, error)
+	getTags() (string, error)
+	getMachineType() (string, error)
+	getLabels() (map[string]string, error)
+	getMetadata() (map[string]string, error)
+	getInterfaceIPv4s() (map[string]string, error)
+}
+
+// List of single-valued attributes (non-nested)
+var singleAttributeSpec = map[gceAttribute]func(gceDataProvider) (string, error){
+	project:      gceDataProvider.getProject,
+	zone:         gceDataProvider.getZone,
+	network:      gceDataProvider.getNetwork,
+	subnetwork:   gceDataProvider.getSubnetwork,
+	publicIP:     gceDataProvider.getPublicIP,
+	privateIP:    gceDataProvider.getPrivateIP,
+	instanceID:   gceDataProvider.getInstanceID,
+	instanceName: gceDataProvider.getInstanceName,
+	tags:         gceDataProvider.getTags,
+	machineType:  gceDataProvider.getMachineType,
+}
+
+// List of nested attributes
+var nestedAttributeSpec = map[gceAttribute]func(gceDataProvider) (map[string]string, error){
+	metadata:      gceDataProvider.getMetadata,
+	interfaceIPv4: gceDataProvider.getInterfaceIPv4s,
+	label:         gceDataProvider.getLabels,
+}
+
+// GCEDetector implements the Detector interface and provide attributes of the VM when on GCE
+type GCEDetector struct {
```

Comment:
[nit] naming feels a bit off to me. `GCEDetector.Project` sounds like the project of the detector and not the metadata of the VM that is being detected. I'm struggling to come up with a better name/interface that works for both GCE and non-GCE resource detection however. Maybe we can just do the simple thing and rename this and `Detector` to `Resource` and the method `GetDetector` to `GetResource`. And we can add a comment that in GCE VM land the resource is the VM metadata.

It is a nit pick tho. Feel free to ignore.

Reply:
I think this is a good idea - I have changed the name to `Resource`.